lib/grpc/client/adapters/mint.ex

```diff
@@ -0,0 +1,227 @@
+    cond do
+      success_bidi_stream?(stream) ->
+        do_receive_data(stream, :bidirectional_stream, opts)
+
+      success_server_stream?(stream) ->
+        do_receive_data(stream, :unary_request_stream_response, opts)
+
+      success_client_stream?(stream) ->
+        do_receive_data(stream, :stream_request_unary_response, opts)
+
+      success_unary_request?(stream) ->
+        do_receive_data(stream, :unary_request_response, opts)
+
+      true ->
+        handle_errors_receive_data(stream, opts)
+    end
+  end
+
+  @impl true
+  def send_headers(%{channel: %{adapter_payload: nil}}, _opts),
+    do: raise("Can't start a client stream without a connection process")
+
+  def send_headers(stream, opts) do
+    do_request(stream, opts, :stream)
+  end
+
+  @impl true
+  def send_data(
+        %{
+          channel: %{adapter_payload: %{conn_pid: pid}},
+          payload: %{response: {:ok, %{request_ref: request_ref}}}
+        } = stream,
+        message,
+        opts
+      ) do
+    {:ok, data, _} = GRPC.Message.to_data(message, opts)
+    :ok = ConnectionProcess.stream_request_body(pid, request_ref, data)
+    if opts[:send_end_stream], do: ConnectionProcess.stream_request_body(pid, request_ref, :eof)
+    stream
+  end
+
+  @impl true
+  def end_stream(
+        %{
+          channel: %{adapter_payload: %{conn_pid: pid}},
+          payload: %{response: {:ok, %{request_ref: request_ref}}}
+        } = stream
+      ) do
+    ConnectionProcess.stream_request_body(pid, request_ref, :eof)
+    stream
+  end
+
+  @impl true
+  def cancel(stream) do
+    %{
+      channel: %{adapter_payload: %{conn_pid: conn_pid}},
+      payload: %{response: {:ok, %{request_ref: request_ref}}}
+    } = stream
+
+    ConnectionProcess.cancel(conn_pid, request_ref)
+  end
+
+  defp connect_opts(%Channel{scheme: "https"} = channel, opts) do
+    %Credential{ssl: ssl} = Map.get(channel, :cred, %Credential{})
+
+    transport_opts =
+      opts
+      |> Keyword.get(:transport_opts, [])
+      |> Keyword.merge(ssl)
+
+    [transport_opts: Keyword.merge(@default_transport_opts, transport_opts)]
+  end
+
+  defp connect_opts(_channel, opts) do
+    transport_opts = Keyword.get(opts, :transport_opts, [])
+    [transport_opts: Keyword.merge(@default_transport_opts, transport_opts)]
+  end
+
+  defp mint_scheme(%Channel{scheme: "https"} = _channel), do: :https
+  defp mint_scheme(_channel), do: :http
+
+  defp do_receive_data(%{payload: %{stream_response_pid: pid}}, request_type, _opts)
+       when request_type in [:bidirectional_stream, :unary_request_stream_response] do
+    stream = StreamResponseProcess.build_stream(pid)
+    {:ok, stream}
+  end
+
+  defp do_receive_data(
+         %{payload: %{stream_response_pid: pid}},
+         request_type,
+         opts
+       )
+       when request_type in [:stream_request_unary_response, :unary_request_response] do
+    with stream <- StreamResponseProcess.build_stream(pid),
```

Comment:
This `with` seems unnecessary because both the first and second clauses always match. This could be refactored into a `case`